lib/BuildSystem/SwiftTools.cpp

```diff
@@ -536,6 +541,50 @@ class SwiftCompilerShellCommand : public ExternalCommand {
     
     return true;
   }
+
+  std::string previousOutput;
+  ssize_t numBytes = 0;
+
+  void partialOutputAvailable(StringRef partialOutput) {
+    previousOutput.append(partialOutput);
+
+    do {
+      if (numBytes == 0) {
+        auto newLinePos = previousOutput.find("\n");
+        // Don't even have the first line yet.
+        if (newLinePos == std::string::npos) {
+          return;
+        }
+
+        StringRef numBytesString = previousOutput.substr(0, newLinePos);
+        if (StringRef(numBytesString).getAsInteger(10, numBytes)) {
+          numBytes = 0;
+          // Couldn't get number of bytes.
+          // Probably due to a compiler crash or bad output?
+          return;
+        }
+
+        // Remove the message length string.
+        previousOutput = previousOutput.substr(newLinePos+1, previousOutput.size());
+      }
+
+      // See if we have enough output to parse.
+      if (previousOutput.size() >= numBytes) {
+        StringRef legibleJSON = previousOutput.substr(0, numBytes);
+
+        // Do something with JSON.
+        std::string sep = "\n--------------------\n";
+        fwrite(sep.data(), sep.size(), 1, stdout);
+        fflush(stdout);
+        fwrite(legibleJSON.data(), legibleJSON.size(), 1, stdout);
+        fflush(stdout);
```

Comment:
When building up incremental patches, I think each patch should be "committable" (i.e., not have behavior which would be problematic if it was landed without the subsequent patches). So, I would factor this patch (on top of first patch to add `hasPartialOutput()` in such a way that it just adds the facility of adding a per-JSON message handler (if enabled).

As you have done here, I would probably test it by just printing the data, but I wouldn't actually commit that. If it was complicated enough I felt like it deserved its own tests then I would design it to have a conditional logging feature the tests could use.